song_scrounger/document_parser.py

```diff
@@ -1,49 +1,128 @@
-def find_quoted_tokens(text):
-    """Retrieves all quoted strings in the order they occur in the given text.
-    Params:
-        text (str).
-
-    Returns:
-        tokens (list): strings found between quotes.
-
-    Notes:
-        - Ignores trailing quote if quotes are unbalanced
-        - Skips empty tokens
-    """
-    if len(text) == 0:
-        return []
-
-    tokens = []
-    while len(text) > 0:
-        quoted_token_indices = _find_first_two_quotes(text)
-        if quoted_token_indices is None:
-            break
-
-        opening_quote_index, closing_quote_index = quoted_token_indices
-        if opening_quote_index < closing_quote_index:
-            tokens.append(text[opening_quote_index+1:closing_quote_index])
-
-        text = "" if closing_quote_index+1 == len(text) else text[closing_quote_index+1:]
-    return tokens
-
-def _find_first_two_quotes(text):
-    """Finds indices of first two quotation marks.
-
-    e.g. 'A "quote"' => (2,8)
-    e.g. 'No quote' => None
-    e.g. 'Not "balanced' => None
-
-    Params:
-        text (str): e.g. 'A "quote"'.
-    Returns:
-        ((int,int)): indices of first two quotes in given text. None if absent or unbalanced.
-    """
-    if len(text) <= 1:
-        return None
-
-    opening_quote_index = text.find("\"")
-    if opening_quote_index != -1:
-        closing_quote_index = text.find("\"", opening_quote_index+1)
-        if closing_quote_index != -1:
-            return opening_quote_index, closing_quote_index
-    return None
\ No newline at end of file
+from .spotify_client import SpotifyClient
+from .models.song import Song
+
+
+class DocumentParser():
+    def __init__(self, spotify_client):
+        self.spotify_client = spotify_client
+
+    async def find_songs(self, text):
+        """Parses given text for songs, matching with artists if mentioned.
+
+        Each song is searched on Spotify. The artists in the search results
+        are searched for in the text as well. Any matches are used for
+        song disambiguation.
+
+        Params:
+            text (str): multi-paragraph page containing song names
+                and perhaps some of their artists.
+
+        Returns:
+            (dict): key (str) is song name; val ([str]) is spotify URIs
+                of matching songs, empty if no matching artist mentioned.
+        """
+        results = {}
+        paragraphs = self._get_paragraphs(text)
+        for paragraph in paragraphs:
+            song_names = self.find_quoted_tokens(paragraph)
+            for song_name in song_names:
+                songs = await self.search_spotify(song_name)
+                songs_with_mentioned_artists = self.filter_by_mentioned_artist(songs, text)
+                if len(songs_with_mentioned_artists) > 0:
```

Comment:
If you define this function
`get_songs_in_text_or_default(self, song_name, text):`
`""" gets spotify songs by song_name and returns a set of those which are in the text.  If none are found in the text, it returns all songs.  """`
`songs = await self.search_spotify(song_name)`
`return filter_by_mentioned_artist(self, songs, text) or songs`

You could write this function in one line
`{ song_name: song.spotify_uri for paragraph in self._get_paragraphs(text) for song_name in self.find_quoted_tokens(paragraph) for song in await self.gets_songs_in_text_or_default(song_name, text) }`

Reply:
I actually pulled out this if block into a separate function. I'll push up the changes shortly.

I personally don't like using complicated comprehensions, I think they start giving diminishing returns.